### PR TITLE
[GR-1320] Display user messages on individual views.

### PR DESCRIPTION
**Pull request by slazicoicr**
It was not possible to inform users of issues with Dashi without a Dashi
release. This solves the problem by allowing a local JSON file to hold
messages displayed on individual views.

### CODE REVIEW

File: application/dash_application/known_pages_router.py [hunk 1]
```diff
@@ -113,3 +148,22 @@ def content_handler(path, qs):
...
+            Output('user_message', 'is_open'),
+        ],
+        [
+            Input('url', 'pathname'),
+        ])
+    def display_user_message(path):
+        """
+        If there is a user message associated with the loaded page, display it.
+        """
+        requested = path[1:] # drop the leading slash
```
Comment (avarsava) [inline application/dash_application/known_pages_router.py:184]: I'm having trouble with this line. It stack traces here with `TypeError: 'NoneType' object is not subscriptable` on every page load, even when my json file is well formed and contains a banner message that displays correctly.

File: README.md [hunk 2]
```diff
@@ -30,6 +30,7 @@ Create a `.env` file in the root directory of this repository:
 | `USE_BLEEDING_EDGE_ETL` | No | Set to install `gsi-qc-etl@master` instead of the release version of `gsi-qc-etl` in `requirements.txt` (Docker only) | `1` | use release version |
 | `EXCLUDE_SWAP_LIBS` | No | File path to TSV file of library pairs to be excluded for swap view | `./exclude_swap_lib.tsv` | |
 | `SAMPLES_FOR_PROJECTS` | No | Indicate whether samples from ALL projects should be used, or only samples from ACTIVE projects. | `ALL` | `ACTIVE` |
+| `DISPLAY_USER_MESSAGE` | No | A JSON file containing a dictionary of page names (key) and messages to display (value)| `./user_messages.json` | |
```
Comment (avarsava) [inline README.md:33]: Could we include a user_messages.json-example like we do for .env?

File: application/dash_application/known_pages_router.py [hunk 3]
```diff
@@ -63,10 +67,41 @@ def menu_item(label, link):
...
 
+def load_user_messages(json_path=os.getenv("DISPLAY_USER_MESSAGE")):
+    if json_path is None:
+        return {}
+
+    if not os.path.isfile(json_path):
+        logger.warning("User message JSON file does not exist")
+        return {}
+
+    # If this fails, it will take down Dashi, so general exception handling is a must
```
Comment (avarsava) [inline application/dash_application/known_pages_router.py]: Should a problem with the message file maybe take down Dashi? Unless we add alerting when these warnings are found in Loki, problems with the message file will fail relatively silently and we might not check that the banner is showing up correctly in prod. If the service crashes then at least we'll hear about it

**Review by avarsava — COMMENTED** [review]
- I would prefer a higher contrast font colour, the brown on yellow seems non-urgent to me
- I would like if we could match the padding and margins of the surrounding elements: 
image

**Comment by avarsava** [comment]
It's possible to make a blank banner by setting eg "single-lane-ts":true

**Comment by avarsava** [comment]
if the same report is used as a key twice, one message is picked for display. Testing is too slow to tell whether it's random or always the bottom one

**Comment by slazicoicr** [comment]
> 
> 
> if the same report is used as a key twice, one message is picked for display. Testing is too slow to tell whether it's random or always the bottom one

From the specs: `The RFC specifies that the names within a JSON object should be unique, but does not mandate how repeated names in JSON objects should be handled. By default, this module does not raise an exception; instead, it ignores all but the last name-value pair for a given name`. Should an exception be raised?

**Comment by avarsava** [comment]
"single-lane-ts":["test1", "test2"]
renders on the page as 
image
I don't know why anyone would do most of these things to be fair

**Comment by avarsava** [comment]
> > if the same report is used as a key twice, one message is picked for display. Testing is too slow to tell whether it's random or always the bottom one
> 
> From the specs: `The RFC specifies that the names within a JSON object should be unique, but does not mandate how repeated names in JSON objects should be handled. By default, this module does not raise an exception; instead, it ignores all but the last name-value pair for a given name`. Should an exception be raised?

If our list ends up growing to a length that's hard to hold entirely in mental memory, throwing an exception in this case might be useful. I think we should avoid having messages getting ignored silently

**Comment by avarsava** [comment]
adding `"secret": "you found the secret page"` actually displays the message on the 404 page for <dashi url>/secret ! I don't know if we should check that the page named actually exists or not. this behaviour might cause entries with typo'd keys to fail silently.

File: application/dash_application/known_pages_router.py (hunk 1 above)
Comment (slazicoicr) [reply application/dash_application/known_pages_router.py:184]: I changed the message to be red.

**Review by asofmosq — APPROVED** [review]

File: application/dash_application/known_pages_router.py [hunk 4]
```diff
@@ -63,10 +67,58 @@ def menu_item(label, link):
...
+
 # Default layout element (wraps the page layout elements which are returned by the router)
 layout = html.Div([
     core.Location(id='url', refresh=False),
     navbar(default_title),
+    dbc.Alert(
+        id="user_message",
+        is_open=False,
+        color="danger",
+        style={"margin-left": "15px"},
```
Comment (avarsava) [inline application/dash_application/known_pages_router.py]: margin-right too please

File: user_messages.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+    "single-lane-rna": "Report is amazing. Just wanted to let you know."
```
Comment (avarsava) [inline user_messages.json:2]: :grin:

**Review by avarsava — APPROVED** [review]

**Pull request merged**